MemoryExtensions.Replace(Span<T>, T, T) implemented

gfoidl opened the pull request:
## Description

Fixes https://github.com/dotnet/runtime/issues/75322

Replaces some open coded loops that I found (or that were linked in the issue), and forwared `string.Replace(char, char)` to this new span-based replace.

## Benchmarks

Run on x64 with AVX2.

### `string.Replace(char, char)`

```
|         Method | Length |      Mean | Ratio |
|--------------- |------- |----------:|------:|
|        Default |      7 |  26.92 ns |  1.00 |
| ViaSpanHelpers |      7 |  23.40 ns |  0.87 |     <- less code in the scalar path
|                |        |           |       |
|        Default |      8 |  29.98 ns |  1.00 |
| ViaSpanHelpers |      8 |  21.30 ns |  0.71 |     <- Vector128<ushort> (= 8 chars) can be used, as opposed to Vector<ushort> (= 16 chars)
|                |        |           |       |
|        Default |     15 |  37.62 ns |  1.00 |
| ViaSpanHelpers |     15 |  22.46 ns |  0.60 |     <- Vector128<ushort> (= 8 chars) can be used, as opposed to Vector<ushort> (= 16 chars)
|                |        |           |       |
|        Default |     16 |  21.11 ns |  1.00 |
| ViaSpanHelpers |     16 |  22.54 ns |  1.07 |     <- the call to the vectorized path, as opposed to vectorized path inlined in string.Replace
|                |        |           |       |
|        Default |     31 |  23.18 ns |  1.00 |
| ViaSpanHelpers |     31 |  23.82 ns |  1.03 |     <- same for the rest, but as longer the length, the less shows the cost of the call
|                |        |           |       |
|        Default |    100 |  33.61 ns |  1.00 |
| ViaSpanHelpers |    100 |  35.31 ns |  1.06 |
|                |        |           |       |
|        Default |    500 |  94.92 ns |  1.00 |
| ViaSpanHelpers |    500 |  96.93 ns |  1.02 |
|                |        |           |       |
|        Default |   1000 | 172.06 ns |  1.00 |
| ViaSpanHelpers |   1000 | 175.28 ns |  1.02 |
```

Is this slight regression OK?
For me yes, as it avoids code duplication and it's a few ns.

### Others

For replacement of the open coded loops I didn't run benchmarks, as in each iteration there is a `call` to `IndexOf`.
Now in the worst case there is one `call` to `ReplaceValueTypeVectorized`.

gfoidl committed:
Defined API

gfoidl committed:
Tests

gfoidl committed:
Scalar implementation

gfoidl committed:
Use EqualityComparer<T>.Default instead

gfoidl committed:
Delegation to SpanHelpers.Replace

gfoidl committed:
ReplaceValueType implemented

gfoidl committed:
Use ushort instead of short, as it doesn't sign-extend for broadcast and in the scalar loop

gfoidl committed:
Forward string.Replace(char, char) to SpanHelpers.ReplaceValueType

gfoidl committed:
Process remainder vectorized only when not done already and with max width available

gfoidl committed:
Split into inlineable scalar path and non-inlineable vectorized path

gfoidl committed:
Replaced open coded loops with Replace

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2618,6 +2618,101 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ReplaceValueType<T>(ref T src, ref T dst, T oldValue, T newValue, nuint length) where T : struct
```
For `string.Replace(char, char)` to work, the signature has to look like this. Cf. Cf. https://github.com/dotnet/runtime/issues/75322#issuecomment-1252842175

Codegen-wise (at least on windows) the `length` is passed via stack. This method will be inlined, so no problem and below for the vectorized code-path (not inlined) the cost will be amortized. So I think this isn't a problem.

stephentoub replied:
For my edification, how much worse it it if string.Replace is implemented as a memcpy and then an in-place replace?  2x worse? Less than that?  I'm wondering whether the two-arg case is important enough that we should be exposing it publicly, or if the in-place replace is sufficient.  We went with only the two-arg case for endianness reversal.

gfoidl replied:
The change would look like 64c7e8136aad79f074842b3ac3fc95aea78a35f8.

For the benchmarks I just tested the worst-case*, i.e. the first char is a match, so the whole string gets copied and then again from the beginning iterated over to replace -- $O(2n)$ instead of $O(n)$ with the out-of-place approach (current PR).

\* string consisting all of `-`, then `string.Replace('-', '+')`

Numbers look like (AVX2 machine):

```
|     Method | Length |      Mean |    Median | Ratio |
|----------- |------- |----------:|----------:|------:|
| OutOfPlace |      7 |  34.65 ns |  34.10 ns |  1.00 |
|    Inplace |      7 |  34.32 ns |  33.12 ns |  1.00 |
|            |        |           |           |       |
| OutOfPlace |      8 |  26.95 ns |  26.08 ns |  1.00 |
|    Inplace |      8 |  27.53 ns |  27.27 ns |  1.01 |
|            |        |           |           |       |
| OutOfPlace |     15 |  24.33 ns |  24.33 ns |  1.00 |
|    Inplace |     15 |  45.03 ns |  43.91 ns |  1.85 |
|            |        |           |           |       |
| OutOfPlace |     16 |  25.90 ns |  24.98 ns |  1.00 |
|    Inplace |     16 |  32.42 ns |  32.27 ns |  1.16 |
|            |        |           |           |       |
| OutOfPlace |     31 |  26.86 ns |  26.69 ns |  1.00 |
|    Inplace |     31 |  36.23 ns |  36.21 ns |  1.35 |
|            |        |           |           |       |
| OutOfPlace |    100 |  37.64 ns |  37.18 ns |  1.00 |
|    Inplace |    100 |  50.73 ns |  50.67 ns |  1.35 |
|            |        |           |           |       |
| OutOfPlace |    500 | 108.17 ns | 107.71 ns |  1.00 |
|    Inplace |    500 | 131.40 ns | 129.90 ns |  1.22 |
|            |        |           |           |       |
| OutOfPlace |   1000 | 192.27 ns | 191.45 ns |  1.00 |
|    Inplace |   1000 | 223.64 ns | 221.66 ns |  1.19 |
```

For the best-case, i.e. no match, the runtime should be equal, as the IndexOf(oldValue)-scan is the only thing executing.
Then of course there are all variations in between.

PS: numbers are a bit flaky on re-runs, but the given numbers are quite representative (no my machine).

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2618,6 +2618,101 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
+            }
+            else
+            {
+                ReplaceValueTypeVectorized(ref src, ref dst, oldValue, newValue, length);
```
I split the vectorized code-path into an extra method (that won't be inlined), so that the scalar-path is tight and fast. Benchmarking for string.Replace showed that this is needed to be on par.

stephentoub replied:
Does this mean the scalar loop is going to be inlined into every MemoryExtensions.Replace call site?  That doesn't seem ideal.

gfoidl replied:
Yes and yes, so I changed (back) to a non-annoted method without the split. 
Benchmark for string.Replace looks now like:
```
|         Method | Length |      Mean |    Median | Ratio |
|--------------- |------- |----------:|----------:|------:|
|        Default |      7 |  29.04 ns |  28.99 ns |  1.00 |
| ViaSpanHelpers |      7 |  28.90 ns |  28.50 ns |  0.94 |
|                |        |           |           |       |
|        Default |      8 |  30.63 ns |  28.76 ns |  1.00 |
| ViaSpanHelpers |      8 |  21.92 ns |  21.76 ns |  0.75 |
|                |        |           |           |       |
|        Default |     15 |  39.29 ns |  39.15 ns |  1.00 |
| ViaSpanHelpers |     15 |  25.06 ns |  24.82 ns |  0.64 |
|                |        |           |           |       |
|        Default |     16 |  25.84 ns |  24.10 ns |  1.00 |
| ViaSpanHelpers |     16 |  27.48 ns |  27.29 ns |  1.09 |
|                |        |           |           |       |
|        Default |     31 |  26.17 ns |  25.56 ns |  1.00 |
| ViaSpanHelpers |     31 |  29.72 ns |  29.31 ns |  1.12 |
|                |        |           |           |       |
|        Default |    100 |  47.76 ns |  46.54 ns |  1.00 |
| ViaSpanHelpers |    100 |  39.16 ns |  38.36 ns |  0.80 |
|                |        |           |           |       |
|        Default |    500 | 108.94 ns | 105.87 ns |  1.00 |
| ViaSpanHelpers |    500 | 119.06 ns | 110.75 ns |  1.10 |
|                |        |           |           |       |
|        Default |   1000 | 206.77 ns | 195.29 ns |  1.00 |
| ViaSpanHelpers |   1000 | 187.44 ns | 184.15 ns |  0.92 |
```
Numbers especially for higher length are flaky. Run multiple times, turned memory-randomizaiton on and off. Sometimes it looks like shown, sometimes e.g. length 1000 is slower.

I don't like that split into methods for scalar / vectorized neither -- it's not natural.
So I pushed the change and then lets see what perf-lab says?

gfoidl reviewed (commented):
Some notes for review.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2618,6 +2618,101 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
+                    val = newValue;
+                }
+            }
+        }
```
This `Replace<T>` should probably be implemented in terms of `IEquatable<T>` rather than `EqualityComparer<T>.Default`, to help with perf for reference types.  It will mean special-casing null, though.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2618,6 +2618,101 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
+                    idx += (uint)Vector128<T>.Count;
+                } while (idx <= lastVectorIndex);
+
+                if (idx < length)
```
Previously you argued against such a guard check.  You measured and changed your mind?

gfoidl replied:
This is really hard to measure, as in micro-benchmarks the branch-predictor will make good decisions.
I'm glad you picked this one up, as I just have my view on this and would like to see some other thoughts or a confirmation.

My take is as follows:
[Previously](https://github.com/dotnet/runtime/pull/67049#discussion_r946253469) we had `Vector<ushort>`, so on AVX that's 16 elements. Thus the probability that no remainder needs to be done is $\frac{1}{16}$ (6.25 %), or in other words: with the guard we penalize $\frac{15}{16}$ (93.75 %) of inputs as the remainders needs to be processed.
Additionally the vector read should be in the cache (as just read a few instructions before, leaving context-switching aside), and an additional store should be handled by cpu's store buffers.

Now we have `Vector128<T>` / `Vector256<T>`, so the probability of not having a remainder changes depending on the size of T.
| size | elements Vec128 | p for remainder Vec 128 | elements Vec256 | p for remainder Vec 256 |
| -- | -- | -- | -- | -- |
| 1 | 16 | $\frac{15}{16}$ (93.75 %) | 32 | $\frac{31}{32}$ (96.875 %) |
| 2 | 8 | $\frac{7}{8}$ (87.5 %) | 16 | $\frac{14}{16}$ (93.75 %) |
| 4 | 4 | $\frac{3}{4}$ (75 %) | 8 | $\frac{7}{8}$ (87.5 %) |
| 8 | 2 | $\frac{1}{2}$ (50 %) | 4 | $\frac{3}{4}$ (75 %) |

So based on that one can argue that the branch-predictor could naively always take the branch (i.e. process the remainder) as this is the most likely case.
Thus the guard is just overhead (fetch, decode, bigger code-size).

In the rare case that the remainder doesn't need to be processed, and the branch-predictor took the branch a pipeline stall will need to happen, costing about 20 cycles.
Just executing the code cost 13 cycles according [godbolt](https://godbolt.org/z/b48G85Tbd).

Sure this is rather naive approach to the branch-predictor, but based on that I'm leaning towards removing that guard for both Vector128 and Vector256, i.e. process the remainder anyway as it's cheaper based on that theory.

<sub>Note: this is of course no rule of thumb, applies only to this use case.</sub>

gfoidl replied:
Ey, in the whole handling of remainders (of idempotent) operations I thought quite a bit, but not far enough...(well mentally caught in the history and scheme that vec-loops _must_ process as much as possible)
Sometimes it can be so easy to just remove a character from the code 😉 
-> ebcbb9fa1e25b0661b150f562b3b398039cac13e (as seen in other recent PRs)

gfoidl committed:
Don't use EqualityComparer<T>.Default

Cf. https://github.com/dotnet/runtime/pull/76337#discussion_r982886319

gfoidl committed:
Remove guards for remainder

Cf. https://github.com/dotnet/runtime/pull/76337#discussion_r983448480

gfoidl committed:
Don't split method into scalar and vectorized and don't force inlining of scalar-part

gfoidl committed:
Merge branch 'main' into memoryextensions_replace

gfoidl commented:
I see failures from CI like
```
JIT/opt/Vectorization/UnrollEqualsStartsWIth/UnrollEqualsStartsWIth.sh
Process terminated. Assertion failed.
   at System.SpanHelpers.ReplaceValueType[T](T& src, T& dst, T oldValue, T newValue, UIntPtr length)
   at System.String.Replace(Char oldChar, Char newChar)
   at UnrollEqualsStartsWIth.RunTests(Type type)
   at UnrollEqualsStartsWIth.Main()
```
This wasn't on my table -- will address them tomorrow.

gfoidl committed:
Fixed assert

ReplaceValueType is called from string.Replace(char, char) so the Debug.Assert was on wrong position, as at entry to method non accelerated platforms are allowed to call it.

gfoidl committed:
Better handling of remainder from the vectorized loop(s)

Intentionally leave one iteration off, as the remaining elements are done vectorized anyway. This eliminates the less probable case (cf. https://github.com/dotnet/runtime/pull/76337#discussion_r983448480) that the last vector is done twice.

stephentoub commented on src/libraries/System.Memory/tests/Span/Replace.T.cs:
```diff
@@ -0,0 +1,148 @@
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.SpanTests;
```
Nit: I don't believe we use file-scoped namespaces anywhere else in this project.  Let's remain consistent.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:
```diff
@@ -2645,6 +2645,68 @@ public static void Sort<T>(this Span<T> span, Comparison<T> comparison)
+                }
+                else if (Unsafe.SizeOf<T>() == sizeof(ushort))
+                {
+                    // Use ushort rather than short, as this avoid a sign-extending move.
```
Should we be using the unsigned types in general in these SpanHelper delegations across all of MemoryExtensions, unless there's a good reason to do otherwise?

gfoidl replied:
I don't remember exactly why, but anywhere in the JIT-dasm I spotted the `movsxd` and that got eliminated by using `ushort`.

Besides this one only the IndexOfAnyInRange-variants use `ushort`.
One would need to review all other places, not that regression may occur. Shall I do that? (If so, I can look into this not earlier than next week).

gfoidl replied:
FYI [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8ovLrl5hcAbho1iAZiakGAYQYBvGg11NlANRhgM0cqQAcAHlwALaBgB8DALIAKOw4YA3AJQMAXmcjEzNLFjVYbAwYNz85aj19BhDTKHNrDk8oJ1c3LPscn38glOM0jIiomLjfBIBfIA=).
Note the difference
```diff
- L0003: movsx rax, r8w
+ L0003: movzx eax, r8w
```
This is simple repo, but I remember to have seen this in the actual code too (and I don't like the `movsx`s if not needed.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2618,6 +2618,101 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
+                        result.StoreUnsafe(ref dst, idx);
+
+                        idx += (uint)Vector128<T>.Count;
+                    } while (idx < lastVectorIndex);
```
Nit:
```suggestion
                    }
                    while (idx < lastVectorIndex);
```

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2618,6 +2618,101 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
+                        result.StoreUnsafe(ref dst, idx);
+
+                        idx += (uint)Vector256<T>.Count;
+                    } while (idx < lastVectorIndex);
```
Nit:
```suggestion
                    }
                    while (idx < lastVectorIndex);
```

stephentoub reviewed (approved):
Thanks!

gfoidl committed:
Merge branch 'main' into memoryextensions_replace

gfoidl committed:
PR feedback

gfoidl commented:
In CI there are lots of
```
2022-11-07T19:29:54.8813391Z   Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/49a1bb2b-12b0-475f-adbd-1560fc76be38/nuget/v3/flat2/system.text.encoding.extensions/index.json'.
2022-11-07T19:29:54.8815846Z   Response status code does not indicate success: 500 (Internal Server Error - Request was blocked due to exceeding usage of resource 'Concurrency' in namespace 'IPAddress'. For more information on why your request was blocked, see the topic "Rate limits" on the Microsoft Web site (https://go.microsoft.com/fwlink/?LinkId=823950). (DevOps Activity ID: 98AA3880-47F6-42F4-923A-307431AC4A03)).
2022-11-07T19:29:59.6386982Z   Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/flat2/system.reflection.extensions/index.json'.
...
2022-11-07T19:32:12.8485509Z   Response status code does not indicate success: 503 (Service Unavailable - TF10216: Azure DevOps services are currently unavailable. Try again later. Activity Id: e149fcd3-05f1-4295-9808-801bb4d7b9cf (DevOps Activity ID: E149FCD3-05F1-4295-9808-801BB4D7B9CF)).
2022-11-07T19:32:14.7782231Z   Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/a65e5cb4-26c0-410f-9457-06db3c5254be/nuget/v3/flat2/runtime.linux-arm64.microsoft.netcore.runtime.jit.tools/index.json'.
2022-11-07T19:32:14.7785140Z   Response status code does not indicate success: 503 (Service Unavailable - TF10216: Azure DevOps services are currently unavailable. Try again later. Activity Id: 33ad97cf-a40e-4562-85fa-49f8be050393 (DevOps Activity ID: 33AD97CF-A40E-4562-85FA-49F8BE050393)).
2022-11-07T19:32:14.9826360Z   Failed to download package 'System.Runtime.4.1.0' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/system.runtime/4.1.0/system.runtime.4.1.0.nupkg'.
...
```

adamsitnik commented:
> In CI there are lots of

I've re-triggered the CI, let's see if it helps.
